src/node_webstorage.cc

```diff
@@ -0,0 +1,661 @@
+  Local<Value> exception;
+
+  if (!domexception_ctor->NewInstance(context, arraysize(argv), argv)
+           .ToLocal(&exception)) {
+    return;
+  }
+
+  isolate->ThrowException(exception);
+}
+
+Storage::Storage(
+    Environment* env, Local<Object> object, Local<String> location) :
+    BaseObject(env, object) {
+  MakeWeak();
+  node::Utf8Value utf8_location(env->isolate(), location);
+  symbols_.Reset(env->isolate(), Map::New(env->isolate()));
+  db_ = nullptr;
+  location_ = utf8_location.ToString();
+}
+
+Storage::~Storage() {
+  CHECK_EQ(sqlite3_close(db_), SQLITE_OK);
+  db_ = nullptr;
+}
+
+void Storage::MemoryInfo(MemoryTracker* tracker) const {
+  tracker->TrackField("symbols", symbols_);
+  tracker->TrackField("location", location_);
+}
+
+bool Storage::Open() {
+  static const char sql[] =
+      "PRAGMA busy_timeout = 3000;"
+      "PRAGMA journal_mode = WAL;"
+      "PRAGMA optimize;"
+      "PRAGMA synchronous = NORMAL;"
+      "PRAGMA temp_store = memory;"
+      ""
+      "CREATE TABLE IF NOT EXISTS nodejs_webstorage("
+      "  key BLOB NOT NULL,"
+      "  value BLOB NOT NULL,"
+      "  size INTEGER NOT NULL,"
+      "  PRIMARY KEY(key)"
+      ") STRICT;"
+      ""
+      "CREATE TABLE IF NOT EXISTS nodejs_webstorage_size("
+      // max_size is 10MB. This can be made configurable in the future.
+      "  max_size INTEGER NOT NULL DEFAULT 10485760,"
+      "  total_size INTEGER NOT NULL,"
+      "  single_row_ INTEGER NOT NULL DEFAULT 1 CHECK(single_row_ = 1),"
+      "  PRIMARY KEY(single_row_)"
+      ") STRICT;"
+      ""
+      "CREATE TRIGGER IF NOT EXISTS nodejs_quota_insert "
+      "AFTER INSERT ON nodejs_webstorage "
+      "FOR EACH ROW "
+      "BEGIN "
+      "  UPDATE nodejs_webstorage_size SET total_size = total_size + NEW.size;"
+      "  SELECT RAISE(ABORT, 'QuotaExceeded')"
+      "  WHERE EXISTS ("
+      "    SELECT 1 FROM nodejs_webstorage_size WHERE total_size > max_size"
+      "  );"
+      "END;"
+      ""
+      "CREATE TRIGGER IF NOT EXISTS nodejs_quota_update "
+      "AFTER UPDATE ON nodejs_webstorage "
+      "FOR EACH ROW "
+      "BEGIN "
+      "  UPDATE nodejs_webstorage_size SET total_size = total_size + "
+      "    (NEW.size - OLD.size);"
+      "  SELECT RAISE(ABORT, 'QuotaExceeded')"
+      "  WHERE EXISTS ("
+      "    SELECT 1 FROM nodejs_webstorage_size WHERE total_size > max_size"
+      "  );"
+      "END;"
+      ""
+      "CREATE TRIGGER IF NOT EXISTS nodejs_quota_delete "
+      "AFTER DELETE ON nodejs_webstorage "
+      "FOR EACH ROW "
+      "BEGIN "
+      "  UPDATE nodejs_webstorage_size SET total_size = total_size - OLD.size;"
+      "END;"
+      ""
+      "INSERT OR IGNORE INTO nodejs_webstorage_size (total_size) VALUES (0)";
+
+  if (db_ != nullptr) {
+    return true;
+  }
+
+  int r = sqlite3_open(location_.c_str(), &db_);
+  CHECK_ERROR_OR_THROW(env(), r, SQLITE_OK, false);
+  r = sqlite3_exec(db_, sql, 0, 0, nullptr);
```

Comment:
> It's a common practice to rely on `IF NOT EXISTS` in cases like this. Instead of checking the `user_version`, a typical database migration would create the table (with `IF NOT EXISTS`) and then perform a series of `ALTER TABLE` operations.

FWIW, in SQLite, `ALTER TABLE` is much more constrained than in PostgreSQL or MySQL or so. Many kinds of changes require making a copy of an entire table. In any case, this requires detecting whether or not the schema needs to be migrated, and I think `user_version` is a good choice for that. (And makes `IF NOT EXISTS` redundant, but I don't feel strongly about that.)